### PR TITLE
Fixed to dump the ignored_scope in quantization more gracefully 

### LINKED ISSUE #2564: [Good First Issue][NNCF]: dump the ignored scope more gracefully
### Context

NNCF saves parameters of the compression to [OpenVINO IR](https://docs.openvino.ai/2024/documentation/openvino-ir-format.html).
It's done in [dump_parameters](https://github.com/openvinotoolkit/nncf/blob/edd397b01f8c45b976785fa67d425467d80835ae/nncf/openvino/rt_info.py#L20) method using [set_rt_info](https://docs.openvino.ai/2024/api/ie_python_api/_autosummary/openvino.runtime.Model.html#openvino.runtime.Model.set_rt_info) API. 
But the problem is that ignored scope is saved not gracefully - it prints all fields even if they are empty. 
image

### What needs to be done?

1) Dump ignored scope in more compact way.

- not dump `<validate value="True" />`
- not dump fields `names`, `patterns` or `types` when they are empty. 
image
- when all fields empty - dump in one line: `<ignored_scope value=[]>`
image

2) Add very lightweight unit test checking the content of <rt_info> for a dummy synthetic model, for example via [get_rt_info](https://docs.openvino.ai/2024/api/ie_python_api/_autosummary/openvino.runtime.Model.html#openvino.runtime.Model.get_rt_info) API. 

### Example Pull Requests

_No response_

### Resources

- [Contribution guide - start here!](https://github.com/openvinotoolkit/nncf/blob/develop/CONTRIBUTING.md)

### Contact points

@ljaljushkin 

### Ticket

129593

### PR DESCRIPTION
**Pull request by Viditagarwal7479**
### Changes
In the case of dumping parameters of `IgnoredScope` in quantization:
- The field `validate` isn't dumped
- If the values corresponding to the rest of the fields are empty (`[]`) they aren't dumped
- In case the `IgnoredScope` is passed blank, then only `<ignored_scope value="[]">` is dumped

### Related tickets
Ticket: 129593
Fixes Issue: #2564 

### Tests

The test case, [`tests/openvino/native/quantization/test_quantization_pipeline.py::test_meta_information`](https://github.com/openvinotoolkit/nncf/blob/86ea8f68636b3fec84339c2cdc1969227f647f6e/tests/openvino/native/quantization/test_quantization_pipeline.py#L97) has been modified to validate the new changes. To protect code from regressions.

### CODE REVIEW

**Comment by Viditagarwal7479** [comment]
Hello @ljaljushkin, a gentle reminder to review the PR.

File: tests/openvino/native/quantization/test_quantization_pipeline.py [hunk 1]
```diff
@@ -103,6 +103,9 @@ def check_parameters(quantized_model, parameters, path):
             if isinstance(value, IgnoredScope):
                 check_parameters(quantized_model, value.__dict__, rt_path)
                 continue
+            if "ignored_scope" in rt_path:
+                assert key != "validate"  # key validate shouldn't be dumped
+                assert value  # empty values shouldn't be dumped
```
Comment (ljaljushkin) [inline tests/openvino/native/quantization/test_quantization_pipeline.py:108]: It doesn't check the content of actual dumped runtime info, instead it directly accesses the input parameters. 
Please add a unit test checking the content of <rt_info> for a dummy synthetic model, for example via get_rt_info.

File: nncf/openvino/rt_info.py [hunk 2]
```diff
@@ -17,6 +17,22 @@
...
 
+def handle_ignored_scope(value: IgnoredScope) -> IgnoredScope:
+    """
+    Dump Ignored Scope parameters more gracefully.
+    https://github.com/openvinotoolkit/nncf/issues/2564
+    Remove keys where value is empty and key is `validate`
+
+    :param value: IgnoredScope instance, based on the quantization parameters
+    """
+    value.__dict__.pop("validate")
```
Comment (ljaljushkin) [inline nncf/openvino/rt_info.py:28]: I'd say that it's not expected from the user standpoint that after passing parameters to `dump_parameters` the parameters are modified. Please use a copy of `IgnoredScope`

**Review by ljaljushkin — CHANGES_REQUESTED** [review]
Thanks for the contribution! 
Please address the comments

**Comment by ljaljushkin** [comment]
@Viditagarwal7479 please fix the test to pass CI

https://github.com/openvinotoolkit/nncf/actions/runs/8344527642/job/22869598068?pr=2578
```
>           assert quantized_model.get_rt_info(rt_path) == str(value)
E           RuntimeError: Check 'rt_info.find(name) != rt_info.end()' failed at src/core/include/openvino/core/model.hpp:502:
E           Cannot get runtime attribute. Path to runtime attribute is incorrect.

tests/openvino/native/quantization/test_quantization_pipeline.py:108: RuntimeError
```

File: tests/openvino/native/quantization/test_quantization_pipeline.py [hunk 3]
```diff
@@ -120,3 +122,49 @@ def check_parameters(quantized_model, parameters, path):
...
+        "subset_size": 1,
+        "fast_bias_correction": True,
+        "ignored_scope": ignored_options,
+    }
+    quantized_model = quantize_impl(model, dataset, **quantize_parameters)
+    ov.save_model(quantized_model, f"{dumped_model_path}")
+    core = ov.Core()
+    dumped_model = core.read_model(f"{dumped_model_path}")
+    os.remove(f"{dumped_model_path}")
+    if id == 1:
```
Comment (ljaljushkin) [inline tests/openvino/native/quantization/test_quantization_pipeline.py:157]: Ideally, unit test should as easy is possible without `if` and `for`. Otherwise, it may catch many errors that are hard to debug, and it's hard to add new test cases. 
Propose refactoring test in the following way:
1. remove id 1,2,3
2. each test case is defined by a pair of parameters: `IgnoredScope` that is dumped, and reference ignored scope (dict) that should be returned by `get_rt_info(["nncf", "quantization", "ignored_scope"])`
3. do only one assert that compares reference dict with actual one.

File: tests/openvino/native/quantization/test_quantization_pipeline.py [hunk 4]
```diff
@@ -120,3 +122,49 @@ def check_parameters(quantized_model, parameters, path):
     assert quantized_model.has_rt_info(base_path)
 
     check_parameters(quantized_model, quantize_parameters, base_path)
+
+
+@pytest.mark.parametrize(
+    "id, ignored_options",
+    zip(
```
Comment (ljaljushkin) [inline tests/openvino/native/quantization/test_quantization_pipeline.py:129]: You can follow this way of defining parameters for each test case without `zip` and `ids`:
https://github.com/openvinotoolkit/nncf/blob/develop/tests/openvino/native/quantization/test_channel_alignment.py#L67-L74

please also refer to the suggestion about refactoring below, these 2 comments are connected.

**Review by ljaljushkin — CHANGES_REQUESTED** [review]

File: tests/openvino/native/quantization/test_quantization_pipeline.py (hunk 3 above)
Comment (Viditagarwal7479) [reply tests/openvino/native/quantization/test_quantization_pipeline.py:157]: 2. `get_rt_info(["nncf", "quantization", "ignored_scope"])` returns value of type `<OVAny class>`, which I guess can't be directly compared with dict, that's why in other tests also they use `for`, `recursion` and `if` statements to compare its values.

File: nncf/openvino/rt_info.py (hunk 2 above)
Comment (Viditagarwal7479) [reply nncf/openvino/rt_info.py]: In this, I created a copy of `ignored_scope` and then made the changes but still, when it goes into `model.set_rt_info` as it dumps the value, it also updates the original data. That newly copied `ignored_scope` value is passed `set_rt_info` here.

**Comment by Viditagarwal7479** [comment]
> @Viditagarwal7479 please fix the test to pass CI
> 
> https://github.com/openvinotoolkit/nncf/actions/runs/8344527642/job/22869598068?pr=2578
> 
> ```
> >           assert quantized_model.get_rt_info(rt_path) == str(value)
> E           RuntimeError: Check 'rt_info.find(name) != rt_info.end()' failed at src/core/include/openvino/core/model.hpp:502:
> E           Cannot get runtime attribute. Path to runtime attribute is incorrect.
> 
> tests/openvino/native/quantization/test_quantization_pipeline.py:108: RuntimeError
> ```

While making modifications through the `model.set_rt_info` we are making changes in the output of `model.get_rt_info` so, I guess we need to update this test case accordingly to facilitate the changes.

File: tests/openvino/native/quantization/test_quantization_pipeline.py [hunk 5]
```diff
@@ -120,3 +121,47 @@ def check_parameters(quantized_model, parameters, path):
...
+                "subgraphs": None,
+                "patterns": None,
+                "names": "['Transpose']",
+            },
+        ),
+        (IgnoredScope(), {"": "[]"}),
+    ],
+)
+def test_ignored_scope_dump(ignored_options, expected_dump, tmp_path):
+    ignored_scpoe_path = ["nncf", "quantization", "ignored_scope"]
```
Comment (ljaljushkin) [inline tests/openvino/native/quantization/test_quantization_pipeline.py:147]: typo
```suggestion
    ignored_scope_path = ["nncf", "quantization", "ignored_scope"]
```

File: tests/openvino/native/quantization/test_quantization_pipeline.py (hunk 3 above)
Comment (ljaljushkin) [reply tests/openvino/native/quantization/test_quantization_pipeline.py]: I see, looks better after refactoring

File: tests/openvino/native/quantization/test_quantization_pipeline.py [hunk 6]
```diff
@@ -120,3 +121,47 @@ def check_parameters(quantized_model, parameters, path):
...
+    dataset = get_dataset_for_test(model)
+    quantize_parameters = {
+        "preset": QuantizationPreset.PERFORMANCE,
+        "target_device": TargetDevice.CPU,
+        "subset_size": 1,
+        "fast_bias_correction": True,
+        "ignored_scope": ignored_options,
+    }
+    quantized_model = quantize_impl(model, dataset, **quantize_parameters)
+    ov.save_model(quantized_model, f"{tmp_path}_ov_model.xml")
```
Comment (ljaljushkin) [inline tests/openvino/native/quantization/test_quantization_pipeline.py:159]: image
pytest prepared a separate folder for each test case, but the model is saved to a common directory. 
I'd do that:
```suggestion
    ov.save_model(quantized_model, tmp_path / "ov_model.xml")
```

**Review by ljaljushkin — CHANGES_REQUESTED** [review]

File: nncf/openvino/rt_info.py (hunk 2 above)
Comment (ljaljushkin) [reply nncf/openvino/rt_info.py]: Could you elaborate this `when it goes into model.set_rt_info as it dumps the value, it also updates the original data`?
Do you mean that still with your latest changes `ignored_scope` got changed on dumping parameters?

File: tests/openvino/native/quantization/test_quantization_pipeline.py (hunk 5 above)
Comment (Viditagarwal7479) [reply tests/openvino/native/quantization/test_quantization_pipeline.py]: Uh! sorry for this silly typos 😅

File: nncf/openvino/rt_info.py (hunk 2 above)
Comment (Viditagarwal7479) [reply nncf/openvino/rt_info.py]: Yes. This is why this [test](https://github.com/openvinotoolkit/nncf/actions/runs/8344527642/job/22869598068?pr=2578) is failing. Since [here](https://github.com/Viditagarwal7479/nncf/blob/3024c889b36de846339200942ecd9cc36f671a23/nncf/openvino/rt_info.py#L62) whatever the value is passed into the `model.set_rt_info` function gets dumped finally and gets updated at the model side which `model.get_rt_info` retrieves.

File: tests/openvino/native/quantization/test_quantization_pipeline.py [hunk 7]
```diff
@@ -120,3 +121,47 @@ def check_parameters(quantized_model, parameters, path):
...
+    dataset = get_dataset_for_test(model)
+    quantize_parameters = {
+        "preset": QuantizationPreset.PERFORMANCE,
+        "target_device": TargetDevice.CPU,
+        "subset_size": 1,
+        "fast_bias_correction": True,
+        "ignored_scope": ignored_options,
+    }
+    quantized_model = quantize_impl(model, dataset, **quantize_parameters)
+    ov.save_model(quantized_model, tmp_path+"/ov_model.xml")
```
Comment (ljaljushkin) [inline tests/openvino/native/quantization/test_quantization_pipeline.py:159]: Please fix the code style
https://github.com/openvinotoolkit/nncf/actions/runs/8361803692/job/22891746357?pr=2578
you can make sure locally that `make pre-commit` is passing

Comment (Viditagarwal7479) [reply tests/openvino/native/quantization/test_quantization_pipeline.py]: Ah, don't know what's happening today. Lots of silly misses!

**Comment by Viditagarwal7479** [comment]
The [`pre-commit-linters / pre-commit (pull_request)`](https://github.com/openvinotoolkit/nncf/actions/runs/8363658310/job/22899096486?pr=2578) is failing because of this unknown reason:
```
Run actions/setup-python@v3
Version 3.8.10 was not found in the local cache
Error: read ECONNRESET
```

**Comment by AlexanderDokuchaev** [comment]
@Viditagarwal7479 rerun the job helped, tests passed.

**Comment by Viditagarwal7479** [comment]
It's showing "1 review requesting changes by reviewers with write access" but I have incorporated all the requested changes so far.

**Review by ljaljushkin — APPROVED** [review]

File: nncf/openvino/rt_info.py [hunk 8]
```diff
@@ -17,6 +17,23 @@
 from nncf.scopes import IgnoredScope
 
 
+def handle_ignored_scope(value: IgnoredScope) -> IgnoredScope:
```
Comment (andrey-churkin) [inline nncf/openvino/rt_info.py:20]: We can return here `Dict[str, Any]` instead of `IgnoredScope`, because we use `__dict__` after that

```suggestion
def handle_ignored_scope(value: IgnoredScope) -> Dict[str, Any]:
```

File: nncf/openvino/rt_info.py [hunk 9]
```diff
@@ -17,6 +17,23 @@
...
+    """
+    Dump Ignored Scope parameters more gracefully.
+    https://github.com/openvinotoolkit/nncf/issues/2564
```
Comment (andrey-churkin) [inline nncf/openvino/rt_info.py:23]: This information is not related to the documentation of function.
```suggestion
```

File: nncf/openvino/rt_info.py (hunk 8 above)
Comment (andrey-churkin) [inline nncf/openvino/rt_info.py:20]: It would be great to have here less general name. For example, the `exclude_empty_fields()` name gives us some information about what this function will do.

File: nncf/openvino/rt_info.py [hunk 10]
```diff
@@ -17,6 +17,23 @@
...
+    Remove keys where value is empty and key is `validate`
+
+    :param value: IgnoredScope instance, based on the quantization parameters
+    """
+    new_value = IgnoredScope(**(value.__dict__))
```
Comment (andrey-churkin) [inline nncf/openvino/rt_info.py:28]: Think about how we can use `dataclasses.asdict` method here.
https://docs.python.org/3/library/dataclasses.html#dataclasses.asdict

**Review by andrey-churkin — CHANGES_REQUESTED** [review]

File: tests/openvino/native/quantization/test_quantization_pipeline.py [hunk 11]
```diff
@@ -120,3 +127,47 @@ def check_parameters(quantized_model, parameters, path):
     assert quantized_model.has_rt_info(base_path)
 
     check_parameters(quantized_model, quantize_parameters, base_path)
+
+
+@pytest.mark.parametrize(
```
[the remark this answers is not in the document]
Comment (Viditagarwal7479) [reply tests/openvino/native/quantization/test_quantization_pipeline.py:132]: Can you please direct me on how to check the subgraphs in a model while quantization? As for names and types, I can check through the dump for the value of these keys.

[the remark this answers is not in the document]
Comment (ljaljushkin) [reply tests/openvino/native/quantization/test_quantization_pipeline.py:132]: @Viditagarwal7479, subgraphs should be dumped as names and types to `rt_info` and checked similarly. 
no need to check how ignored scope with subgraphs is applied on quantization, if you asked about that.

[the remark this answers is not in the document]
Comment (Viditagarwal7479) [reply tests/openvino/native/quantization/test_quantization_pipeline.py:132]: No, I meant that how should I know the name of the subgraph present in any model and how a subgraph passed into IgnoredScope? As there's no reference in the documentation related to this.

[the remark this answers is not in the document]
Comment (Viditagarwal7479) [reply tests/openvino/native/quantization/test_quantization_pipeline.py:132]: Like for `names` I know by going through the documentation that it should be a list of string and how to get the value of the passed string. For example `["Transpose"]`

**Comment by ljaljushkin** [comment]
@Viditagarwal7479, the last commit was a month ago. There's still one unresolved comment. Are you going to resolve it? Do you need a help? Please inform us if you do not plan to continue working on this task. Thanks!

**Comment by Viditagarwal7479** [comment]
> @Viditagarwal7479, the last commit was a month ago. There's still one unresolved comment. Are you going to resolve it? Do you need a help? Please inform us if you do not plan to continue working on this task. Thanks!

Yes, I need help in determining the parameters for `subgraph` in `IgnoredScope` for a given model. I looked at the PR you mentioned, but that PR wasn't of much help in resolving this, one thing that can be of help is if we also use the same model used in the PR, then as requested, we can include a `subgraph` based test case for this PR.

**Comment by ljaljushkin** [comment]
> > @Viditagarwal7479, the last commit was a month ago. There's still one unresolved comment. Are you going to resolve it? Do you need a help? Please inform us if you do not plan to continue working on this task. Thanks!
> 
> Yes, I need help in determining the parameters for `subgraph` in `IgnoredScope` for a given model. I looked at the PR you mentioned, but that PR wasn't of much help in resolving this, one thing that can be of help is if we also use the same model used in the PR, then as requested, we can include a `subgraph` based test case for this PR.

Firstly, I suggest using not quantized model, since the `ignored scope` defines nodes that should not be quantized. 
If they are already quantized, this attribute does not make sense and do nothing.

For instance, you can consider WeightsModel.
image
To not quantize all layers in the subgraph marked red, you can specify the names of input and output nodes, as follows.
```python
IgnoredScope(
    subgraphs=[
        nncf.Subgraph(
            inputs=[
                "MatMul_1",
            ],
            outputs=[
                "MatMul"
            ],
        )
    ],
),
```
Hope this example, makes things clear.

**Comment by ljaljushkin** [comment]
@Viditagarwal7479 one week left since my last comment
Please inform us if you do not plan to continue working on this task. Thanks!

**Comment by Viditagarwal7479** [comment]
> @Viditagarwal7479 one week left since my last comment Please inform us if you do not plan to continue working on this task. Thanks!

Thank you so much, @ljaljushkin, for helping me (in fact, spoon-feeding on how to add the subgraph-based test case). 

Through the diagram, you showed it's clear, but just to make sure, for any model whose subgraph we want to ignore in the quantised model, we just have to mention the `input node` and the `output node`. Through this, the nodes that are taking the input as the value passed from the mentioned `input node` and whose output is, in a way, being passed to the mentioned `output node` get removed from the quantised model. Is my interpretation correct?

Sorry, for the delay from my end in making the minor changes. If my above interpretation is correct so should we make some more guides related to how to specify the subgraph in the IgnoredScope maybe just by adding similar subgraph based test cases or a tutorial kind of stuff?

**Comment by ljaljushkin** [comment]
> > @Viditagarwal7479 one week left since my last comment Please inform us if you do not plan to continue working on this task. Thanks!
> 
> Thank you so much, @ljaljushkin, for helping me (in fact, spoon-feeding on how to add the subgraph-based test case).
> 
> Through the diagram, you showed it's clear, but just to make sure, for any model whose subgraph we want to ignore in the quantised model, we just have to mention the `input node` and the `output node`. Through this, the nodes that are taking the input as the value passed from the mentioned `input node` and whose output is, in a way, being passed to the mentioned `output node` get removed from the quantised model. Is my interpretation correct?
> 
> Sorry, for the delay from my end in making the minor changes. If my above interpretation is correct so should we make some more guides related to how to specify the subgraph in the IgnoredScope maybe just by adding similar subgraph based test cases or a tutorial kind of stuff?

Yes, your interpretation is correct. More guidance about defining subgraph in the IgnoredScope would be definitely useful. 
Feel free to add it in another PR. Probably, just updating docstring with some examples for IgnoredScope class can be enough. 
You can visualize my example using symbols and tabs:
```
  |                  |     
input1        input2
   \               /
    node...
        |
       ...
        |
     output1
```

**Comment by Viditagarwal7479** [comment]
> > > @Viditagarwal7479 one week left since my last comment Please inform us if you do not plan to continue working on this task. Thanks!
> > 
> > 
> > Thank you so much, @ljaljushkin, for helping me (in fact, spoon-feeding on how to add the subgraph-based test case).
> > Through the diagram, you showed it's clear, but just to make sure, for any model whose subgraph we want to ignore in the quantised model, we just have to mention the `input node` and the `output node`. Through this, the nodes that are taking the input as the value passed from the mentioned `input node` and whose output is, in a way, being passed to the mentioned `output node` get removed from the quantised model. Is my interpretation correct?
> > Sorry, for the delay from my end in making the minor changes. If my above interpretation is correct so should we make some more guides related to how to specify the subgraph in the IgnoredScope maybe just by adding similar subgraph based test cases or a tutorial kind of stuff?
> 
> Yes, your interpretation is correct. More guidance about defining subgraph in the IgnoredScope would be definitely useful. Feel free to add it in another PR. Probably, just updating docstring with some examples for IgnoredScope class can be enough. You can visualize my example using symbols and tabs:
> 
> ```
>   |                  |     
> input1        input2
>    \               /
>     node...
>         |
>        ...
>         |
>      output1
> ```

Thank you for the confirmation. I will soon (2-3 days) create a PR corresponding to this.

**Review by ljaljushkin — APPROVED** [review]

**Review by andrey-churkin — APPROVED** [review]
